pkg/controllers/etcd.go

```diff
@@ -103,18 +103,26 @@ func (s *EtcdService) Run(ctx context.Context, ready chan<- struct{}, stopped ch
 		return fmt.Errorf("%s failed to start: %v", s.Name(), err)
 	}
 
-	// Make sure microshift-etcd is terminated whether it failed to properly start or MicroShift stops
+	// Handle microshift-etcd termination before microshift process exits
+	go func() {
+		if err := cmd.Wait(); err != nil {
+			klog.Warningf("%v failed waiting on process to finish: %+v", s.Name(), err)
+		}
+		klog.Infof("%v process quit: %v", s.Name(), cmd.ProcessState.String())
+
+		// Exit microshift to trigger microshift-etcd restart
+		klog.Warning("microshift-etcd process terminated prematurely, restarting MicroShift")
+		os.Exit(0)
+	}()
+
+	// Handle microshift-etcd termination after microshift process exits
 	defer func() {
 		if err := cmd.Process.Signal(os.Interrupt); err != nil {
 			klog.Warningf("%v failed interrupting the process: %+v", s.Name(), err)
 			if err := cmd.Process.Kill(); err != nil {
 				klog.Warningf("%v failed killing the process: %+v", s.Name(), err)
 			}
 		}
-		if err := cmd.Wait(); err != nil {
```

Comment:
Shouldn't we wait until microshift-etcd finishes? I hope that etcd shutdown sequence won't be interrupted

Reply:
On Linux, the process management is using fork->exec->wait pattern, which we need to mimic in Go.
Wait is a blocking function, which only exits after the process terminates. 
It's the way to detect and properly "release" an orphan process in case `microshift-etcd` exits prematurely.

Reply:
But now you have no guarantee that cmd.Wait from line 108 will return, go doesn't wait for goroutines by itself. Is there a risk that we signal etcd to finish, return from this function and almost immediately whole `microshift` exits not giving chance that `goroutine` to wait for etcd to finish?

Reply:
If microshift exits first without running Wait on microshift-etcd, the orphaned process is cleaned up by the `init` (pid 1) OS process. 
You can simulate this situation by running `kill -9` on microshift - the OS would take care of the clean up in this case.

Reply:
So problem we're solving here is detecting microshift-etcd exit and restarting to get it up again? If there are no more caveats to this solution, then it looks good enough to me but I'd like others to take a look